{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 1: Controls\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n",
    "\n",
    "# Answers:\n",
    "First of all, it is neccesary to import the csv library to be able to use its functions, with the command **\"import csv\"**. With this, we can handle the input files (\"LocusGene.tsv\" and \"Germplasm.tsv\").\n",
    "\n",
    "The first two variables we declared (\"sameorder\" and \"mismatch\") are used to create the output. The former serves as a **counter of the line that is being checked in each iteration of the loops** and the latter serves as a **counter of the times two lines are not the same in both documents**.\n",
    "\n",
    "Next, we have to open the files in *read mode* --> using open (\"file\", \"r\"). Then, using the csv.DictReader function, we load the data from both files to their respective functions (\"locus_data\" and \"germplasm_data\"). Since these are \".tsv\" files, the delimiter is set to a tab (\"\\t\"). Even though there are no quotation marks in the documents, I set it to double quotation marks (\") just in case of need in the future.\n",
    "\n",
    "Then, all we have to do is a ***for*** loop. The for loop nested structure in this code works because when the files are read, the file pointer is set to the end of the last line read. With each iteration of the loops, the next lines are read and the pointer is subsecuentially placed on the end of those lines. Also, since we intruduced a ***\"break\"*** in the nested loop, only one iteration is run, resulting in the designed behaviour consisting of **comparing the first line in the LocusGene file with the first line in the Germplasm file, the second with the second and so on.**\n",
    "\n",
    "As for what the loops do themselves, they simply **print a string stating the line number that is being compared, show the comparison *per se* and showing the result of the comparisons thanks to the if/else block**. When I say that the lines from both documents are being compared, I mean that the \"Locus\" column from both documents is being compared. This is done in the **\" if row1[\"Locus\"] == row2[\"Locus\"]\"** line. That is because the **DictReader function outputs a variant of a dictionary created from separating each column of the document**. Its keys are the header line's text for each column (if present, otherwise they can be declared in the DictReader statement) and the values are, well, the values of each row corresponding to each column (in this case, for example, the first element of the dictionary would be {\"Locus\" : \"AT1G01040\"}, the second {\"Gene\" : \"DCL1\"} and so on until the first line is recorded. Then the loop would iterate over the second line and fill in the values in that row). If the lines are not the same, the **\"mismatch\"** counter is incremented. At the end of each iteration the **\"sameorder\"** counter increments as well to show the next lines to be checked. The \"sameorder\" function is set to 1 at the beginning because even though technically the first lines to be compared are number 0 (at least in Python), it is easier for a human to understand if the count begins in 1.\n",
    "\n",
    "Finally, the final results are created. If the \"mismatch\" variable's value is 0, that means that **all comparisons showed that the lines were the same, so both documents are in the same order.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now checking line number: 1\n",
      "AT1G01040 vs. AT1G01040\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 2\n",
      "AT1G01060 vs. AT1G01060\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 3\n",
      "AT1G01140 vs. AT1G01140\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 4\n",
      "AT1G01220 vs. AT1G01220\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 5\n",
      "AT2G03720 vs. AT2G03720\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 6\n",
      "AT2G03800 vs. AT2G03800\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 7\n",
      "AT2G04240 vs. AT2G04240\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 8\n",
      "AT2G05210 vs. AT2G05210\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 9\n",
      "AT3G02130 vs. AT3G02130\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 10\n",
      "AT3G02140 vs. AT3G02140\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 11\n",
      "AT3G02230 vs. AT3G02230\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 12\n",
      "AT3G02260 vs. AT3G02260\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 13\n",
      "AT3G02310 vs. AT3G02310\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 14\n",
      "AT3G02680 vs. AT3G02680\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 15\n",
      "AT3G02850 vs. AT3G02850\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 16\n",
      "AT3G02870 vs. AT3G02870\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 17\n",
      "AT3G03260 vs. AT3G03260\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 18\n",
      "AT4G14790 vs. AT4G14790\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 19\n",
      "AT4G15210 vs. AT4G15210\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 20\n",
      "AT4G15560 vs. AT4G15560\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 21\n",
      "AT4G15570 vs. AT4G15570\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 22\n",
      "AT4G15802 vs. AT4G15802\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 23\n",
      "AT4G15880 vs. AT4G15880\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 24\n",
      "AT4G16420 vs. AT4G16420\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 25\n",
      "AT4G16480 vs. AT4G16480\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 26\n",
      "AT5G10480 vs. AT5G10480\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 27\n",
      "AT5G10510 vs. AT5G10510\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 28\n",
      "AT5G11110 vs. AT5G11110\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 29\n",
      "AT5G11260 vs. AT5G11260\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 30\n",
      "AT5G11510 vs. AT5G11510\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 31\n",
      "AT5G12200 vs. AT5G12200\n",
      "Result: both are the same\n",
      "\n",
      "Now checking line number: 32\n",
      "AT5G13290 vs. AT5G13290\n",
      "Result: both are the same\n",
      "\n",
      "FINAL RESULTS:\n",
      "All lines are in the same order!\n"
     ]
    }
   ],
   "source": [
    "import csv #First of all, we need to import the csv library to be able to use its functions\n",
    "\n",
    "#These variables are used in creating the output of the file.\n",
    "sameorder = 1\n",
    "mismatch = 0\n",
    "\n",
    "#Next, we need to open both files in read mode and read their contents with the csv.DictReader function\n",
    "germplasm_file =  open (\"Germplasm.tsv\", \"r\")\n",
    "germplasm_data = csv.DictReader(germplasm_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "locus_file = open(\"LocusGene.tsv\", \"r\")\n",
    "locus_data = csv.DictReader(locus_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "#print(germplasm_data)\n",
    "#print (locus_data)\n",
    "\n",
    "#The nested for loops allows us to easily compare both files.\n",
    "for row1 in locus_data:   \n",
    "    for row2 in germplasm_data:\n",
    "        print(\"Now checking line number: {}\".format(sameorder))\n",
    "        print(row1[\"Locus\"] + \" vs. \" + row2[\"Locus\"])\n",
    "        if row1[\"Locus\"] == row2[\"Locus\"]:\n",
    "            print(\"Result: both are the same\\n\")\n",
    "        else:\n",
    "            print(\"Result: both are not the same\\n\")\n",
    "            mismatch += 1\n",
    "        sameorder +=1    \n",
    "        break\n",
    "        \n",
    "#The final results are created here.        \n",
    "print (\"FINAL RESULTS:\")                \n",
    "if mismatch == 0:\n",
    "    print(\"All lines are in the same order!\")\n",
    "else:\n",
    "    print(\"Warning! {} lines are not in the same order\".format(mismatch))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 2: Design and create the database\n",
    "\n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n",
    "# Answers:\n",
    "In this particular case, I finally decided to use **sqlMagic**, due to (from my point of view) its output is much easier to read.\n",
    "\n",
    "First, I had to run the code that is used to connect to MySQL in Docker. All these previous steps are explained in the code box. It is important to have in mind all these previous steps, because otherwise, we would not be able to move on to the following ones (errors of different kinds can occur).\n",
    "\n",
    "Then I had to create a database for this exam, called **\"genetics\"**, and inside of that database I created two tables (**\"germplasm\"** and **\"locus_gene\"**). Since **the \"Locus\" column is shared between both input files (Germplasm.tsv and LocusGene.tsv)** and locus' codes shouldn't appear more than once, I decided to use them as the ***primary keys*** for their respective tables. The locus column is the primary key, which means that **its values must be unique in the entire table**. It also means that **this will be the column that will be used to link both tables in this case**. This will be no problem, since as it has been proven in the last problem, **the AGI Locus Code in both input files are in the same sequence.**\n",
    "The values contained in this column (\"Locus\") must be **VARCHAR**, variable characters **up to length 15** (even though all codes in the input file are 9 characters long, we are simply ensuring that if there is any AGI code that is longer it can fit). The \"pubmed\" column is to be filled with INTEGERs (as well as the \"protein_length\" column in the locus_gene table). The \"NOT NULL\" command ensures us that all columns must contain information (basically, they can't be null). Both tables have different columns but a similar configuration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sql extension is already loaded. To reload it, use:\n",
      "  %reload_ext sql\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1007, \"Can't create database 'genetics'; database exists\")\n",
      "[SQL: CREATE DATABASE genetics]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_genetics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus_gene</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('germplasm',), ('locus_gene',)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##Previous steps --> just to check in the terminal that we have the docker images needed:\n",
    "#docker images\n",
    "#docker start course-mysql\n",
    "#mysql -h 127.0.0.1 -P 3306 --protocol=tcp -u root -p  ## password: \"root\"\n",
    "\n",
    "#These commands enable us to to connect to MySQL in Docker\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql          \n",
    "#%sql show databases\n",
    "\n",
    "#Now I will create my own database, which will contain the two tables of interest\n",
    "%sql CREATE DATABASE genetics\n",
    "%sql USE genetics\n",
    "#%sql show databases\n",
    "\n",
    "\n",
    "#%sql DROP TABLE germplasm; #Just in case we need to delete the table and start over.\n",
    "#%sql DROP TABLE locus_gene; #Just in case we need to delete the table and start over.\n",
    "\n",
    "#These commands create the new table \"germplasm\"\n",
    "%sql CREATE TABLE germplasm(locus VARCHAR(15) PRIMARY KEY NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(2000) NOT NULL, pubmed INTEGER NOT NULL)\n",
    "%sql DESCRIBE germplasm    #Just checking that the table has been created properly\n",
    "\n",
    "#These commands create the new table \"locus_gene\"\n",
    "%sql CREATE TABLE locus_gene(locus VARCHAR(15) PRIMARY KEY NOT NULL, gene VARCHAR(30) NOT NULL, protein_length INTEGER NOT NULL);\n",
    "%sql DESCRIBE locus_gene   #Just checking that the table has been created properly\n",
    "\n",
    "#%sql SELECT * FROM germplasm   #Another way of checking that the table has been created properly\n",
    "#%sql SELECT * FROM locus_gene  #Another way of checking that the table has been created properly\n",
    "%sql show tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(2000)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(2000)', 'NO', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql DESCRIBE germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(15)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>gene</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>protein_length</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(15)', 'NO', 'PRI', None, ''),\n",
       " ('gene', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('protein_length', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql DESCRIBE locus_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n",
    "\n",
    "# Answers:\n",
    "As previously stated, right now the pointer in both files is at their end. Before doing anything else with them, it is necessary to ***reset the pointers' position to the beginning of the first data line***. We can do this by setting them at the beginning of the file and then reading the first line, the header line, skipping it and setting the pointers in the desired position.\n",
    "\n",
    "Once again, we have to iterate over both files. I did this with a for loop for each one. Both of them are essentially the same. As I said before, \"row\" is, in this case, a variation of dictionary that contains the name of each column as keys and the values corresponding to each entry for those columns as values. In the SQL commands I'm inserting into each of the database's columns their corresponding values by accessing the values in the row dictionary and adding them in the correct place. The **.format** method allows me to introduce \"placeholders\" in the SQL command ***({})*** and then passing the corresponding Python variables to each placeholder in order of appearance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors   #This command imports the pymysql.cursors library\n",
    "\n",
    "#Set the pointer to the beginning of the first line of data (skipping the header line)\n",
    "germplasm_file.seek(0)\n",
    "germplasm_file.readline()\n",
    "locus_file.seek(0)\n",
    "locus_file.readline()\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "#insert the data\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for row in locus_data:\n",
    "            sql = \"\"\"INSERT INTO locus_gene (locus, gene, protein_length)\n",
    "            VALUES ('{}', '{}', {});\"\"\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "            cursor.execute(sql) \n",
    "        for row in germplasm_data:\n",
    "            sql = \"\"\"INSERT INTO germplasm (locus, germplasm, phenotype, pubmed)\n",
    "            VALUES ('{}', '{}', '{}', {});\"\"\".format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "            cursor.execute(sql) ### Maybe this script does not comply with the DRY principle, but I couldn't think of anything better\n",
    "    connection.commit() #this commits the changes to the database.\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() #Here I'm closing the connection to the database.\n",
    "    locus_file.close() #Here I'm closing the connection to the \"LocusGene.tsv\" file.\n",
    "    germplasm_file.close() #Here I'm closing the connection to the \"Germplasm.tsv\" file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm #Just checking that the data was properly inserted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>protein_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locus_gene #Just checking that the data was properly inserted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 4: Create reports, written to a file\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish.\n",
    "\n",
    "# Answers:\n",
    "\n",
    "For the ***Report 1***, first of all, as we previously already did, it is neccesary to import the csv library to be able to use its functions, with the command **\"import csv\"**; and also **stablishing a connection with the database** (with connection = pymysql.connect([...])). After that, we will use the **try/finally** commands, to create the desired report from our database. **These three steps are common for the 4 different Reports.** The next step will be similar in each report, having some differences in each case. In this report, after connecting with the recent created database \"genetics\", we retrieved the different columns of the tables with **\"SELECT\"**, selected the tables with **\"FROM\"** and indicated with **\"WHERE\"** that the column \"locus\" is the same in both tables. All of this must be inside a **\"try\"** command. With the following commands we are finally gathering all the columns previouly stated from the database to a new single output --> **results1**. This will show the full, joined, content of the two database tables (including a header line). After that, the last thing to do is (after the **\"finally\"** command) ending the connection with the database (connection.close()), to ensure everything works as intended --> **IT HAS TO BE DONE WITH ALL THE REPORTS**.\n",
    "**As we can see, both files were in the same order. The first three columns correspond to the 'LocusGene.tsv' file and the last four correspond to the file 'Germplasm.tsv'.**\n",
    "\n",
    "For the ***Report 2***, we will just have to add the opperation asked (\"Create a joined report that only includes the Genes SKOR and MAA3\") with the one previously created in Report 1. I did this by creating the own Report 2 operation inside the \"with\" command: '''AND gene in ('SKOR', 'MAA3')'''. After that, combining this with the one of the Report 1 --> **cursor.execute(sql1 + sql2)** makes that **it only shows the records that contains the Genes SKOR and MAA3**.\n",
    "Final result: **Joined tables filtered by Genes 'SKOR' and 'MAA3'.**\n",
    "\n",
    "For the ***Report 3***, as we need to retrieve the information separately inside each different chromosome, it can be done this way: \"for i in range(1, 6):\" --> we are indicating that there will be 5 possibilities: [1, 2, 3, 4, 5]. These numbers will be retrieved from this code: \"sql3 = f\"SELECT COUNT(*) AS NumberOfEntriesChr{i} FROM germplasm WHERE locus LIKE 'AT{i}G%'\"\", which tells us that the number of the records from the table \"germplasm\" containing the structure **'AT{i}G%'** will be counted separately, according to the third character of this structure. So, what we are interested in inside this structure is the **third character** (represented as \"i\"), which will give us the number of the chromosome (as we already know, between 1 and 5) of each record. With the \"for\" command, we iterate this operation throughout the entire database.\n",
    "Final result: **As we can see, there are 4 entries for chromosome 1, 4 entries for chromosome 2, 9 entries for chromosome 3, 8 entries for chromosome 4 and 7 entries for chromosome 5.**\n",
    "\n",
    "For the ***Report 4***, the procedure is quite similar to the one carried out in report 3. The difference here is that the table needed is \"locus_gene\", and instead of counting the different records in each chromosome, **we need to retrieve the average protein length for the genes on each Chromosome**. This can be done with this code: sql4 = f\"SELECT AVG(protein_length) AS avgLengthChr{i} FROM locus_gene WHERE locus LIKE 'AT{i}G%'\".\n",
    "**As we can see, chromosome 1 has an average protein length of 258.75, chromosome 2 has an average protein length of 215.5, chromosome 3 has an average protein length of 252, chromosome 4 has an average protein length of 277.5 and finally chromosome 5 has an average protein length of 271.2857.**\n",
    "\n",
    "\n",
    "Finally, the last thing to do is ***to write all the reports created in one same file.***\n",
    "Here, the main command that is going to be needed is the **write function**. We need to create and define the file that will contain all our reports. In our case is --> **'Reports_Sergio.tsv'**.\n",
    "\n",
    "The main issue will be operating with all these four reports. As we have previously seen, **Reports 1 and 2 are quite similar, and the same goes for Reports 3 and 4**. Thus, I have decided to iterate with a \"for\" loop with the first two reports, and with another \"for\" loop with the other two (note that in the first \"for\", we indicate that the code starts with the value = 1 --> first report; and in the second \"for\", the code starts with the value = 3 --> Third report). So, the main structure of these two iterations will be quite similar, taking into account that the main difference is defined by the **results of the reports themselves** --> [results1 and results2 in the first \"for\"] and [results3 and results4 in the second \"for\"]; and the **start value** (previously commented).\n",
    "\n",
    "The last (but not least) thing that we must have in mind is to stablish the writemode into ***append flag -->'a'*** at the end of each of the two \"for\" loops. **This will allow us to open the file and add new information to it, without destroying the existing information.**\n",
    "\n",
    "**It is also important to check that all reports have been written properly in the file that we just created** (view last code)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors #This command imports the pymysql.cursors library\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql1 = '''\n",
    "                SELECT germplasm.locus, gene, protein_length, germplasm, phenotype, pubmed\n",
    "                FROM germplasm, locus_gene \n",
    "                WHERE germplasm.locus = locus_gene.locus\n",
    "              '''\n",
    "        cursor.execute(sql1)\n",
    "        results1 = cursor.fetchall()\n",
    "finally:\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'locus': 'AT1G01040',\n",
       "  'gene': 'DCL1',\n",
       "  'protein_length': 332,\n",
       "  'germplasm': 'CS3828',\n",
       "  'phenotype': 'Increased abundance of miRNA precursors.',\n",
       "  'pubmed': 17369351},\n",
       " {'locus': 'AT1G01060',\n",
       "  'gene': 'LHY',\n",
       "  'protein_length': 290,\n",
       "  'germplasm': 'lhy-101',\n",
       "  'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.',\n",
       "  'pubmed': 16891401},\n",
       " {'locus': 'AT1G01140',\n",
       "  'gene': 'CIPK9',\n",
       "  'protein_length': 223,\n",
       "  'germplasm': 'SALK_058629',\n",
       "  'phenotype': 'hypersensitive to low potassium media',\n",
       "  'pubmed': 17486125},\n",
       " {'locus': 'AT1G01220',\n",
       "  'gene': 'FKGP',\n",
       "  'protein_length': 190,\n",
       "  'germplasm': 'SALK_012400C',\n",
       "  'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.',\n",
       "  'pubmed': 18199744},\n",
       " {'locus': 'AT2G03720',\n",
       "  'gene': 'MRH6',\n",
       "  'protein_length': 189,\n",
       "  'germplasm': 'SALK_042433',\n",
       "  'phenotype': 'Multiple straight hairs',\n",
       "  'pubmed': 16367956},\n",
       " {'locus': 'AT2G03800',\n",
       "  'gene': 'GEK1',\n",
       "  'protein_length': 196,\n",
       "  'germplasm': 'gek1-1',\n",
       "  'phenotype': 'Ethanol hypersensitivity.',\n",
       "  'pubmed': 15215505},\n",
       " {'locus': 'AT2G04240',\n",
       "  'gene': 'XERICO',\n",
       "  'protein_length': 256,\n",
       "  'germplasm': 'xerico',\n",
       "  'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.',\n",
       "  'pubmed': 17933900},\n",
       " {'locus': 'AT2G05210',\n",
       "  'gene': 'POT1A',\n",
       "  'protein_length': 221,\n",
       "  'germplasm': 'pot1-1',\n",
       "  'phenotype': 'No visible phenotype.',\n",
       "  'pubmed': 17627276},\n",
       " {'locus': 'AT3G02130',\n",
       "  'gene': 'RPK2',\n",
       "  'protein_length': 284,\n",
       "  'germplasm': 'rpk2-2',\n",
       "  'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.',\n",
       "  'pubmed': 17419837},\n",
       " {'locus': 'AT3G02140',\n",
       "  'gene': 'TMAC2',\n",
       "  'protein_length': 300,\n",
       "  'germplasm': 'afp4-1',\n",
       "  'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.',\n",
       "  'pubmed': 18484180},\n",
       " {'locus': 'AT3G02230',\n",
       "  'gene': 'RGP1',\n",
       "  'protein_length': 301,\n",
       "  'germplasm': 'rgp1-1',\n",
       "  'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.',\n",
       "  'pubmed': 21478444},\n",
       " {'locus': 'AT3G02260',\n",
       "  'gene': 'BIG',\n",
       "  'protein_length': 279,\n",
       "  'germplasm': 'tir3-1 RGLG1:rglg1 rglg2',\n",
       "  'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.',\n",
       "  'pubmed': 17586653},\n",
       " {'locus': 'AT3G02310',\n",
       "  'gene': 'SEP2',\n",
       "  'protein_length': 175,\n",
       "  'germplasm': 'sep2-1',\n",
       "  'phenotype': 'Non-described subtle phenotype.',\n",
       "  'pubmed': 10821278},\n",
       " {'locus': 'AT3G02680',\n",
       "  'gene': 'NBS1',\n",
       "  'protein_length': 190,\n",
       "  'germplasm': 'atnbs1-1',\n",
       "  'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.',\n",
       "  'pubmed': 17672843},\n",
       " {'locus': 'AT3G02850',\n",
       "  'gene': 'SKOR',\n",
       "  'protein_length': 234,\n",
       "  'germplasm': 'CS3816',\n",
       "  'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.',\n",
       "  'pubmed': 17568770},\n",
       " {'locus': 'AT3G02870',\n",
       "  'gene': 'VTC4',\n",
       "  'protein_length': 311,\n",
       "  'germplasm': 'vtc4-1',\n",
       "  'phenotype': 'ascorbate deficient',\n",
       "  'pubmed': 16595667},\n",
       " {'locus': 'AT3G03260',\n",
       "  'gene': 'HDG8',\n",
       "  'protein_length': 194,\n",
       "  'germplasm': 'hdg8-1',\n",
       "  'phenotype': 'No visible phenotype.',\n",
       "  'pubmed': 16778018},\n",
       " {'locus': 'AT4G14790',\n",
       "  'gene': 'SUV3',\n",
       "  'protein_length': 312,\n",
       "  'germplasm': 'pdd17',\n",
       "  'phenotype': 'Defective pollen development.',\n",
       "  'pubmed': 19237690},\n",
       " {'locus': 'AT4G15210',\n",
       "  'gene': 'BAM5',\n",
       "  'protein_length': 313,\n",
       "  'germplasm': 'bmy1-2',\n",
       "  'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.',\n",
       "  'pubmed': 16297066},\n",
       " {'locus': 'AT4G15560',\n",
       "  'gene': 'DXS',\n",
       "  'protein_length': 219,\n",
       "  'germplasm': 'cla1-1',\n",
       "  'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.',\n",
       "  'pubmed': 10982425},\n",
       " {'locus': 'AT4G15570',\n",
       "  'gene': 'MAA3',\n",
       "  'protein_length': 294,\n",
       "  'germplasm': 'maa3',\n",
       "  'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.',\n",
       "  'pubmed': 18772186},\n",
       " {'locus': 'AT4G15802',\n",
       "  'gene': 'HSBP',\n",
       "  'protein_length': 254,\n",
       "  'germplasm': 'Athspb-2',\n",
       "  'phenotype': 'Early flowering, reduced fertility, aborted seeds.',\n",
       "  'pubmed': 20388662},\n",
       " {'locus': 'AT4G15880',\n",
       "  'gene': 'ESD4',\n",
       "  'protein_length': 265,\n",
       "  'germplasm': 'esd4-2',\n",
       "  'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.',\n",
       "  'pubmed': 17513499},\n",
       " {'locus': 'AT4G16420',\n",
       "  'gene': 'ADA2B',\n",
       "  'protein_length': 279,\n",
       "  'germplasm': 'prz1-1',\n",
       "  'phenotype': 'Altered response to auxin and cytokinin',\n",
       "  'pubmed': 12747832},\n",
       " {'locus': 'AT4G16480',\n",
       "  'gene': 'INT4',\n",
       "  'protein_length': 284,\n",
       "  'germplasm': 'atint4-2',\n",
       "  'phenotype': 'No visible phenotype.',\n",
       "  'pubmed': 16603666},\n",
       " {'locus': 'AT5G10480',\n",
       "  'gene': 'PAS2',\n",
       "  'protein_length': 301,\n",
       "  'germplasm': 'pas2-3',\n",
       "  'phenotype': 'Segregates 25% embryo lethal.',\n",
       "  'pubmed': 18799749},\n",
       " {'locus': 'AT5G10510',\n",
       "  'gene': 'AIL6',\n",
       "  'protein_length': 310,\n",
       "  'germplasm': 'plt3-1',\n",
       "  'phenotype': 'Short roots and shortened root meristem.',\n",
       "  'pubmed': 17960244},\n",
       " {'locus': 'AT5G11110',\n",
       "  'gene': 'SPS2',\n",
       "  'protein_length': 232,\n",
       "  'germplasm': 'kns2',\n",
       "  'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.',\n",
       "  'pubmed': 18779216},\n",
       " {'locus': 'AT5G11260',\n",
       "  'gene': 'HY5',\n",
       "  'protein_length': 221,\n",
       "  'germplasm': 'hy5-101',\n",
       "  'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.',\n",
       "  'pubmed': 16891401},\n",
       " {'locus': 'AT5G11510',\n",
       "  'gene': 'MYB3R-4',\n",
       "  'protein_length': 336,\n",
       "  'germplasm': 'myb3r4-1',\n",
       "  'phenotype': 'No visible phenotype.',\n",
       "  'pubmed': 17287251},\n",
       " {'locus': 'AT5G12200',\n",
       "  'gene': 'PYD2',\n",
       "  'protein_length': 310,\n",
       "  'germplasm': 'pyd2-2',\n",
       "  'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.',\n",
       "  'pubmed': 19413687},\n",
       " {'locus': 'AT5G13290',\n",
       "  'gene': 'CRN',\n",
       "  'protein_length': 189,\n",
       "  'germplasm': 'crn-1',\n",
       "  'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.',\n",
       "  'pubmed': 12345678}]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors #This command imports the pymysql.cursors library\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql2 = '''\n",
    "                AND gene in ('SKOR', 'MAA3')\n",
    "              '''\n",
    "        cursor.execute(sql1 + sql2)\n",
    "        results2 = cursor.fetchall()\n",
    "finally:\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'locus': 'AT3G02850',\n",
       "  'gene': 'SKOR',\n",
       "  'protein_length': 234,\n",
       "  'germplasm': 'CS3816',\n",
       "  'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.',\n",
       "  'pubmed': 17568770},\n",
       " {'locus': 'AT4G15570',\n",
       "  'gene': 'MAA3',\n",
       "  'protein_length': 294,\n",
       "  'germplasm': 'maa3',\n",
       "  'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.',\n",
       "  'pubmed': 18772186}]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors #This command imports the pymysql.cursors library\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    results3 = []\n",
    "    with connection.cursor() as cursor:\n",
    "        for i in range(1, 6):\n",
    "            sql3 = f\"SELECT COUNT(*) AS NumberOfEntriesChr{i} FROM germplasm WHERE locus LIKE 'AT{i}G%'\"\n",
    "            cursor.execute(sql3)\n",
    "            results3.extend(cursor.fetchall())\n",
    "finally:\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'NumberOfEntriesChr1': 4},\n",
       " {'NumberOfEntriesChr2': 4},\n",
       " {'NumberOfEntriesChr3': 9},\n",
       " {'NumberOfEntriesChr4': 8},\n",
       " {'NumberOfEntriesChr5': 7}]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors #This command imports the pymysql.cursors library\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    results4 = []\n",
    "    with connection.cursor() as cursor:\n",
    "        for i in range(1, 6):\n",
    "            sql4 = f\"SELECT AVG(protein_length) AS avgLengthChr{i} FROM locus_gene WHERE locus LIKE 'AT{i}G%'\"\n",
    "            cursor.execute(sql4)\n",
    "            results4.extend(cursor.fetchall())\n",
    "finally:\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'avgLengthChr1': Decimal('258.7500')},\n",
       " {'avgLengthChr2': Decimal('215.5000')},\n",
       " {'avgLengthChr3': Decimal('252.0000')},\n",
       " {'avgLengthChr4': Decimal('277.5000')},\n",
       " {'avgLengthChr5': Decimal('271.2857')}]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WRITING REPORTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "writemode = 'w'\n",
    "for i, results in enumerate([results1, results2], start=1):\n",
    "    with open('Reports_Sergio.tsv', writemode) as file:\n",
    "        file.write(f'Report{i}\\n')\n",
    "        file.write('\\t'.join([str(k) for k in results[0].keys()]) + '\\n')\n",
    "        for result in results:\n",
    "            file.write('\\t'.join([str(v) for v in result.values()]) + '\\n')\n",
    "        file.write('\\n\\n')\n",
    "    writemode = 'a'\n",
    "    \n",
    "for i, results in enumerate([results3, results4], start=3):\n",
    "    with open('Reports_Sergio.tsv', writemode) as file:\n",
    "        file.write(f'Report{i}\\n')\n",
    "        for result in results:\n",
    "            file.write('\\t'.join([str(k)+'. '+str(v)  for k,v in result.items()]) + '\\n')\n",
    "        file.write('\\n\\n')\n",
    "    writemode = 'a'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Report1\n",
      "locus\tgene\tprotein_length\tgermplasm\tphenotype\tpubmed\n",
      "AT1G01040\tDCL1\t332\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\n",
      "AT1G01060\tLHY\t290\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\n",
      "AT1G01140\tCIPK9\t223\tSALK_058629\thypersensitive to low potassium media\t17486125\n",
      "AT1G01220\tFKGP\t190\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\n",
      "AT2G03720\tMRH6\t189\tSALK_042433\tMultiple straight hairs\t16367956\n",
      "AT2G03800\tGEK1\t196\tgek1-1\tEthanol hypersensitivity.\t15215505\n",
      "AT2G04240\tXERICO\t256\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\n",
      "AT2G05210\tPOT1A\t221\tpot1-1\tNo visible phenotype.\t17627276\n",
      "AT3G02130\tRPK2\t284\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\n",
      "AT3G02140\tTMAC2\t300\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\n",
      "AT3G02230\tRGP1\t301\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\n",
      "AT3G02260\tBIG\t279\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\n",
      "AT3G02310\tSEP2\t175\tsep2-1\tNon-described subtle phenotype.\t10821278\n",
      "AT3G02680\tNBS1\t190\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\n",
      "AT3G02850\tSKOR\t234\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\n",
      "AT3G02870\tVTC4\t311\tvtc4-1\tascorbate deficient\t16595667\n",
      "AT3G03260\tHDG8\t194\thdg8-1\tNo visible phenotype.\t16778018\n",
      "AT4G14790\tSUV3\t312\tpdd17\tDefective pollen development.\t19237690\n",
      "AT4G15210\tBAM5\t313\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\n",
      "AT4G15560\tDXS\t219\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\n",
      "AT4G15570\tMAA3\t294\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\n",
      "AT4G15802\tHSBP\t254\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\n",
      "AT4G15880\tESD4\t265\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\n",
      "AT4G16420\tADA2B\t279\tprz1-1\tAltered response to auxin and cytokinin\t12747832\n",
      "AT4G16480\tINT4\t284\tatint4-2\tNo visible phenotype.\t16603666\n",
      "AT5G10480\tPAS2\t301\tpas2-3\tSegregates 25% embryo lethal.\t18799749\n",
      "AT5G10510\tAIL6\t310\tplt3-1\tShort roots and shortened root meristem.\t17960244\n",
      "AT5G11110\tSPS2\t232\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\n",
      "AT5G11260\tHY5\t221\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\n",
      "AT5G11510\tMYB3R-4\t336\tmyb3r4-1\tNo visible phenotype.\t17287251\n",
      "AT5G12200\tPYD2\t310\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\n",
      "AT5G13290\tCRN\t189\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\n",
      "\n",
      "\n",
      "Report2\n",
      "locus\tgene\tprotein_length\tgermplasm\tphenotype\tpubmed\n",
      "AT3G02850\tSKOR\t234\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\n",
      "AT4G15570\tMAA3\t294\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\n",
      "\n",
      "\n",
      "Report3\n",
      "NumberOfEntriesChr1. 4\n",
      "NumberOfEntriesChr2. 4\n",
      "NumberOfEntriesChr3. 9\n",
      "NumberOfEntriesChr4. 8\n",
      "NumberOfEntriesChr5. 7\n",
      "\n",
      "\n",
      "Report4\n",
      "avgLengthChr1. 258.7500\n",
      "avgLengthChr2. 215.5000\n",
      "avgLengthChr3. 252.0000\n",
      "avgLengthChr4. 277.5000\n",
      "avgLengthChr5. 271.2857\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "reportsfile = open(\"Reports_Sergio.tsv\", \"r\")  #Just checking that all reports have been written properly in the file that we just created\n",
    "print(reportsfile.read())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
